Track thread needs_attention state; shortcut to view threads needing attention

liewegas opened the pull request:
- new table `threads` tracks needs_attention state
- new shortcut shows threads for the current user needing attention
- sql statement to create the table after the fact, and populate it with historic data.  (Note that this doesn't account for #lobby and #demo-lobby... those should be manually set to needs_attention=false.)
- message shortcuts to set/clear needs_attention

TODO
- [ ] daily reminder
- [x] slash command to summarize needs_attention threads in a channel
- [x] operator comamnds

liewegas committed:
progress

liewegas committed:
only unset attention if voter is claimed

liewegas committed:
prettier

liewegas committed:
add message/thread commands to set/clear needs_attention

liewegas committed:
asdf

liewegas committed:
shortcut to query needs_attention

liewegas committed:
table creation

also, populate the historical threads.

note that this does not handle the #lobby and #demo-lobby properly, or
any thread that we routed away from.  :/

liewegas committed:
const

liewegas committed:
remove stray includes

liewegas committed:
fix set/clear needs_attention so it works on any msg in the thread

liewegas committed:
/unclaimed in a channel

benweissmann commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+    , channel_id
+    , user_id
+    , user_phone_number
+    , CASE WHEN direction='INBOUND' THEN true ELSE false END as needs_attention
```
Unassigned threads should be marked as needs attention, even if the last message was outbound

benweissmann commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+--- Table to explicit track thread state
+
+CREATE TABLE threads (
+    slack_parent_message_ts text,
```
You should add indexes to all of these columns (except maybe user_phone_number?)

benweissmann commented on src/db_api_util.ts:
```diff
@@ -591,6 +756,22 @@ export async function logVolunteerVoterClaimToDb(
+  const client = await pool.connect();
+  try {
+    const result = await client.query(
+      'SELECT * FROM volunteer_voter_claims WHERE user_id = $1 AND user_phone_number = $2',
```
This is a bit faster: `SELECT EXISTS(SELECT 1 FROM volunteer_voter_claims WHERE user_id = $1 AND user_phone_number = $2))` (because it's not reading any values, it can figure out the answer straight from the index without having to load in the pages of data)

benweissmann reviewed (commented):
This looks amazing!

tomerovadia commented on external_config/slack.md:
```diff
@@ -81,3 +81,24 @@ This should be added as a **messages** shortcut.
+
+##### Add to ToDo
+
+- Messages shortcut
```
To keep the format consistent, here and below can you match the line in `Reset Demo` (or change it to match what you have here)?

`This should be added as a **messages** shortcut.`

tomerovadia commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+
+CREATE TABLE threads (
+    slack_parent_message_ts text,
+    channel_id text,
```
For consistency across tables can you name this `slack_channel_id`?

tomerovadia commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+    slack_parent_message_ts text,
+    channel_id text,
+    user_id text,
+	user_phone_number text,
```
unindent

tomerovadia commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+	user_phone_number text,
+    needs_attention bool,
+    updated_at timestamp without time zone,
+    PRIMARY KEY (slack_parent_message_ts)
```
I believe you'll want `slack_parent_message_ts` and `slack_channel_id` to both be a multi-column primary key, since `slack_parent_message_ts` is only guaranteed to be unique within a certain channel.

tomerovadia commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+WITH foo AS (
+    SELECT
+        slack_parent_message_ts
+        , slack_channel as channel_id
```
Are there formatting issues with the commas?

tomerovadia commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+    PRIMARY KEY (slack_parent_message_ts)
+);
+
+INSERT INTO threads
```
can you add a comment about what this script does?

liewegas committed:
/unclaimed * or /unclaimed channelname

tomerovadia commented on src/app.ts:
```diff
@@ -607,8 +607,9 @@ app.post(
-      payload.callback_id === 'reset_demo'
+      (payload.type === 'message_action' &&
+        payload.callback_id === 'reset_demo') ||
+      payload.type === 'shortcut'
```
indent

liewegas replied:
this is what prettier did

tomerovadia commented on sql/20201012.sql:
```diff
@@ -0,0 +1,36 @@
+)
+SELECT
+    slack_parent_message_ts
+    , channel_id
```
slack_channel_id for consistency with other tables

tomerovadia commented on src/async_jobs.ts:
```diff
@@ -42,6 +42,25 @@ async function slackInteractivityHandler(
     );
   }
 
+  if (payload.type === 'shortcut') {
```
can you add a comment that this is for global shortcuts?

liewegas committed:
fix tests

tomerovadia commented on src/async_jobs.ts:
```diff
@@ -42,6 +62,25 @@ async function slackInteractivityHandler(
+    }
+    switch (payload.callback_id) {
+      case 'show_needs_attention': {
+        await SlackInteractionHandler.receiveShowNeedsAttention({
```
can you rename to `handleShowNeedsAttention`? Elsewhere I use `receive` to mean receive the command but don't execute it (ask for confirmation), and `handle` is to do the actual execution logic

tomerovadia commented on src/async_jobs.ts:
```diff
@@ -102,6 +141,21 @@ async function slackInteractivityHandler(
           `SERVER POST /slack-interactivity: Determined user interaction is a message shortcut.`
         );
 
+        if (payload.callback_id === 'set_needs_attention') {
```
can you make this a switch statement on the callback_id (and I believe the below logic is callback_id = reset_demo)?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+  age: number | null;
+};
+
+export async function newThreadToDb(
```
can we make this `logThreadToDb` to keep the naming consistent across the file?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+        databaseThreadEntry.needsAttention,
+      ]
+    );
+    logger.info('DBAPIUTIL.newThreadToDb: Successfully created thread');
```
logThreadToDb

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
   failureReason?: string | null;
 };
 
+export type DatabaseThreadEntry = {
```
I have this file ordered such that the functions associated with tables that are most important (e.g. messages) are at top. Can you move these functions to the bottom?

liewegas committed:
make getVoterHasVolunteer faster

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+  const client = await pool.connect();
+  try {
+    await client.query(
+      'UPDATE threads SET needs_attention = $1 WHERE slack_parent_message_ts = $2;',
```
slack_parent_message_ts is only guaranteed to be unique within a channel, so we need to add channel to the WHERE clause here

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+  }
+}
+
+export async function getThreadLatestMessage(
```
This function only gets the message ts, right? Can we rename `getThreadLatestMessageTs`?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+        FROM threads t
+        WHERE t.needs_attention AND t.channel_id = $1
+          AND NOT EXISTS (SELECT FROM volunteer_voter_claims c WHERE t.user_id=c.user_id
+          AND t.user_phone_number=c.user_phone_number)`,
```
can you indent this to make it clear it's part of the subquery?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+         t.slack_parent_message_ts as id, t.channel_id, t.user_phone_number, t.user_id, EXTRACT(EPOCH FROM now() - t.updated_at) as age
+        FROM threads t
+        WHERE t.needs_attention AND t.channel_id = $1
+          AND NOT EXISTS (SELECT FROM volunteer_voter_claims c WHERE t.user_id=c.user_id
```
you shouldn't need to check both the user_id and user_phone_number -- the user_id is derived from the user_phone_number

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+    if (channelId) {
+      result = await client.query(
+        `SELECT
+         t.slack_parent_message_ts as id, t.channel_id, t.user_phone_number, t.user_id, EXTRACT(EPOCH FROM now() - t.updated_at) as age
```
is age the age of the voter or age of the todo? can you make that clearer?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+  }
+}
+
+export async function getThreadNeedsAttentionFor(
```
I see the difference between getThreadNeedsAttentionFor and getThreadsNeedingAttentionFor but can you please add comments since it can be confusing?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -86,6 +86,184 @@ export type DatabaseCommandEntry = {
+  const client = await pool.connect();
+  try {
+    const result = await client.query(
+      'SELECT needs_attention FROM threads WHERE slack_parent_message_ts = $1',
```
You're going to need to filter by both slack_parent_message_ts and slack_channel_id to guarantee you only get one row

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -123,10 +301,37 @@ export async function logMessageToDb(
     logger.info(
       `DBAPIUTIL.logMessageToDb: Successfully inserted message into PostgreSQL database.`
     );
+
```
I'd prefer to keep logMessageToDb's responsibility strictly to logging a message to the DB. Can we add this logic elsewhere, perhaps in a function called alongside where logMessageToDb is called?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -591,6 +796,22 @@ export async function logVolunteerVoterClaimToDb(
+  const client = await pool.connect();
+  try {
+    const result = await client.query(
+      'SELECT * FROM volunteer_voter_claims WHERE user_id = $1 AND user_phone_number = $2',
```
nit: you shouldn't need to check both user_id and user_phone_number. doesn't hurt if you prefer it though. if you choose one go with user_id

tomerovadia commented on src/router.ts:
```diff
@@ -704,6 +731,11 @@ const routeVoterToSlackChannel = async (
     `timestampOfLastMessageInThread: ${timestampOfLastMessageInThread}`
   );
 
+  await DbApiUtil.setThreadNeedsAttentionToDb(
```
This updates the thread to which a voter is returning to have the needsAttention status of the thread from which they are coming from, right? Can you add a comment to that effect?

tomerovadia commented on src/router.ts:
```diff
@@ -632,6 +650,15 @@ const routeVoterToSlackChannel = async (
     // Remember the voter's thread in this channel.
     userInfo[response.data.channel] = response.data.ts;
 
+    // Create the thread
```
can you add to the comment that this new thread will adopt the old thread's needAttention state?

tomerovadia commented on src/slack_api_util.ts:
```diff
@@ -168,6 +203,64 @@ export function copyUserInfoToDbMessageEntry(
+> {
+  const response = await slackAPI.get('conversations.list', {
+    params: {
+      limit: 1000,
```
Slack recommends no more than 200 at a time. We have a function fetchSlackChannelNamesAndIds in this file that already does this with pagination https://api.slack.com/methods/conversations.list

tomerovadia commented on src/slack_api_util.ts:
```diff
@@ -168,6 +203,64 @@ export function copyUserInfoToDbMessageEntry(
+  return [fw, bw];
+}
+
+export async function fetchSlackChannelId(
```
We have a Redis cache to look up a Slack id given a Slack name: RedisApiUtil.slackPodChannelIds

tomerovadia commented on src/slack_api_util.ts:
```diff
@@ -26,13 +26,48 @@ type SlackSendMessageResponse = {
   [channelId: string]: string; // mapping of channel ID to channel name
 };
 
+export async function getThreadPermalink(
```
Can we have the SlackApiUtil function just do the Slack API call and have the DB lookup separate?

liewegas committed:
indexes on threads column

liewegas committed:
fix initial threads population

Unassigned threads are always needs_attention

liewegas committed:
remove dups from todo

liewegas committed:
record chat history ts in threads table

This lets us deep link into the thread even for new conversations

liewegas committed:
update slack command names, formatting

liewegas committed:
channel_id -> slack_channel_id

liewegas committed:
avoid unfurling threads when posting /unclaimed content

liewegas committed:
tab -> spaces

liewegas committed:
newThreadToDb -> logThreadToDb

liewegas committed:
reorder function is db_api_util

liewegas committed:
include channel in PK for threads

liewegas committed:
rename getThreadLatestMessage -> getThreadLatestMessageTs

liewegas committed:
clean up SQL formatting

liewegas committed:
userId is already the md5 of user_phone_number

liewegas committed:
age -> last_update_age

liewegas committed:
comments

liewegas committed:
move thread status update out of logMessageToDb

liewegas committed:
no need to check user_phone_number here too

liewegas committed:
comments

liewegas committed:
use existing redis cache to map channel ids to names and back

liewegas committed:
prettier

liewegas committed:
fix unclaimed listing title

liewegas committed:
separate permalink helper from latest message helper

liewegas committed:
receiveShowNeedsAttention -> handleShowNeedsAttention

liewegas committed:
refactor/simplify slackInteractivityHandler()

I dropped the error path that populates the modal when there is an invalid
shortcut_id or redis error... I don't think it is worth the complexity.

liewegas committed:
prettier

liewegas committed:
/needs-attention command to mirror the shortcut

I find this faster/easier.

liewegas committed:
restrict /unclaimed args (* or channel name) to admins

liewegas committed:
make channel name a link in /unclaimed * output

liewegas committed:
order everything oldest to newest

liewegas committed:
eslint

liewegas committed:
fix needs-attention to query only LATEST voter claim only

liewegas committed:
/needs-attention: accept * or channel name

liewegas committed:
eslint

liewegas committed:
/needs-attention * -- break down by both channel and volunteer

tomerovadia commented on sql/20201012.sql:
```diff
@@ -20,6 +25,7 @@ WITH foo AS (
         , COALESCE(slack_send_timestamp, slack_receive_timestamp) as updated_at
         , direction
         , automated
+        , archived
```
Is `archived` used?

liewegas replied:
removed

tomerovadia commented on sql/20201012.sql:
```diff
@@ -18,24 +18,28 @@ CREATE INDEX ON threads (updated_at);
     SELECT
-        slack_parent_message_ts
+        m.slack_parent_message_ts
         , slack_channel as channel_id
```
can you name this `slack_channel_id` for consistency?

tomerovadia commented on sql/20201012.sql:
```diff
@@ -18,24 +18,28 @@ CREATE INDEX ON threads (updated_at);
     , user_id
     , user_phone_number
-    , CASE WHEN direction='INBOUND' THEN true ELSE false END as needs_attention
+    , CASE WHEN direction='INBOUND' OR volunteer_slack_user_id IS NULL) THEN true ELSE false END as needs_attention
```
Am I interpreting this correctly that this defines `needs_attention` as either a) the last message was from the voter, or b) no volunteer has claimed the voter?

If so, can you add a comment to that effect?

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -212,12 +212,12 @@ export async function getThreadsNeedingAttentionFor(
-          AND c.volunteer_slack_user_id=$1`,
+         slack_parent_message_ts as id, channel_id, user_phone_number, user_id, EXTRACT(EPOCH FROM now() - updated_at) as age
+        FROM threads t
+        WHERE needs_attention AND EXISTS (
```
This is checking whether the user has ever claimed the voter. Did you intend to check whether the voter is currently claimed by the user? If so you'll need to get the latest voter_volunteer_claim to see the latest volunteer who claimed the voter.

liewegas replied:
fixed in later patches

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -145,19 +145,44 @@ export async function setThreadNeedsAttentionToDb(
+}
+
 export async function getThreadLatestMessage(
   slackParentMessageTs: string
```
Is the purpose of `history_ts` to more easily be able to access the latest message in a thread? If so, why do you need to join on messages here -- cant you use `history_ts` instead? Looks like you use both, but does that defeat the purpose of `history_ts`?

liewegas replied:
The history_ts is the id of the 'chat history' message that opens up a thread.  It's used only if there are no other messages in the thread, because otherwise when you link to it slack will highlight the message in the channel and not open up a thread view.  This function will prefer to use the newest message in the chat, otherwise the history item as a fallback.

We could always link to the top of the chat (the history), but I think the latest message in the conversation is better!

tomerovadia commented on external_config/slack.md:
```diff
@@ -92,23 +92,26 @@ This should be added as a **messages** shortcut.
 
-- Global shortcut
-- **Name:** ToDo
+This should be added as a **Global** shortcut.
```
lowercase `global` (or uppercase `messages` above)

liewegas committed:
resolve slack user names

liewegas committed:
make /needs-attention @username work

This isn't very efficient--we should cache this data in redis?

liewegas committed:
pull slack usernames from the database

getSlackUserName() now has no callers, but leaving it for now?

liewegas committed:
drop stray debug line

liewegas committed:
rely on slack escaping instead of looking up usernames ourselves

liewegas committed:
remove dead code

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -385,23 +386,27 @@ export async function handleCommandUnclaimed(
+    slackChannelNames[slackChannelIds[name]] = name;
   }
+
+  logger.info(JSON.stringify(slackChannelIds));
```
This can be an extremely long list. Can we omit, or if you think it's necessary, add some context so when we see it in the logs we know what it's indicating?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -385,23 +386,27 @@ export async function handleCommandUnclaimed(
-  let channelBw: Record<string, string> = {};
-  if (text) {
-    [channelFw, channelBw] = await SlackApiUtil.fetchSlackChannelMap();
+  const slackChannelIds = text
```
If you're using this to get the Slack channel ID from the name, could the slash command escaping option take care of that instead?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -417,9 +417,12 @@ export async function handleCommandUnclaimed(
       (text === '*' ? 'all channels' : `#${channelName}`),
   ];
   for (const x of threads) {
```
Can we replace `x` with `thread` so this block is easier to read?

tomerovadia commented on src/async_jobs.ts:
```diff
@@ -42,66 +78,59 @@ async function slackInteractivityHandler(
+    switch (payload.callback_id) {
+      case 'set_needs_attention': {
+        await DbApiUtil.setThreadNeedsAttentionToDb(
+          payload.message.thread_ts || payload.message.ts,
```
Is there a reason one of these would exist and the other wouldn't? If so, can you add a comment?

tomerovadia commented on src/async_jobs.ts:
```diff
@@ -42,66 +78,59 @@ async function slackInteractivityHandler(
     );
 
     const redisHashKey = `${payload.channel.id}:${payload.message.ts}`;
     const redisData = await RedisApiUtil.getHash(redisClient, redisHashKey);
```
Do you handle a case in which this message command is used on a message that is not part of a still-active voter thread (e.g. no redisData is returned)?

tomerovadia commented on src/async_jobs.ts:
```diff
@@ -126,38 +155,77 @@ async function slackInteractivityHandler(
           "actionTs": payload.action_ts
         } as SlackModalPrivateMetadata;
 
-        if (!originatingSlackChannelName || !redisData) {
```
Do you drop this logic entirely? After your logic, does a feedback modal still appear if a user attempts Reset Demo in an invalid context, or does the loading modal hang in these situations?

liewegas replied:
fixed

tomerovadia commented on external_config/slack.md:
```diff
@@ -81,6 +81,13 @@ Important: The Slack bot must be given access to each of these channels.
 - *Short Description:* List unclaimed voters in a channel
```
Is `Slash Commands` a section under `Interactivity & Shortcuts`? It appears to me as a section of its own on the left.

tomerovadia commented on external_config/slack.md:
```diff
@@ -81,6 +81,13 @@ Important: The Slack bot must be given access to each of these channels.
+- *Command:* /needs-attention
+- *Request URL:* base URL + /slack-command
+- *Short Description:* List voters needing your attention
+- *Usage Hint:*
```
Can you add a usage hint?

liewegas replied:
I'm conflicted... the arguments are only available to admins, not volunteers.

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -457,16 +458,8 @@ export async function handleCommandUnclaimed(
+  const threads = (await DbApiUtil.getThreadsNeedingAttentionFor(userId)) || [];
 
   const urls: string[] = [];
   for (const x of threads) {
```
Can you use `thread` instead of `x`?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -479,20 +472,64 @@ export async function handleShowNeedsAttention({
     urls.push(url);
   }
 
+  const lines = threads.map(
```
Can you remove this use of `.map` and incorporate the logic into the above loop through threads? Especially because map is asynchronous it's not great to have it deconstruct a separate data structure as a side effect.

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -479,20 +472,64 @@ export async function handleShowNeedsAttention({
+  });
+}
+
+export async function handleShowNeedsAttention({
```
Is the difference between `handleCommandNeedsAttention` and `handleShowNeedsAttention` that the former is a slash command and the latter is a global shortcut? If so, can we rename the latter to `handleShortcutNeedsAttention`?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -396,26 +402,26 @@ export async function handleCommandUnclaimed(
+      arg = arg.substr(1);
     }
-    if (!(text in slackChannelIds)) {
+    if (!(arg in slackChannelIds)) {
```
Can we replace much of this logic with the Slack slash command mention escaping feature?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -396,26 +402,26 @@ export async function handleCommandUnclaimed(
-    text === '*' ? null : channelId
+    arg === '*' ? null : channelId
   );
   const lines: string[] = [
```
Am I understanding correctly that these initial values are of information that appears at the top of the message sent in response to the slash command? If so can you add a comment?

liewegas replied:
clearer in latest code

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -383,11 +383,17 @@ export function prettyTimeInterval(seconds: number): string {
   const outputChannelId = channelId;
 
-  const slackChannelIds = text
+  var arg = text;
```
Can you rename arg (perhaps `slashCommandArg`)? Its purpose wasn't clear to me.

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -396,26 +402,26 @@ export async function handleCommandUnclaimed(
-    if (text[0] == '#') {
-      text = text.substr(1);
+  if (arg && arg != '*') {
+    if (arg[0] == '#') {
```
`===` instead of `==`

tomerovadia replied:
I see what you're doing here but can you please add comments so that other readers or us in the future can quickly discern why this is necessary?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -396,26 +402,26 @@ export async function handleCommandUnclaimed(
     }
-    channelName = text;
+    channelName = arg;
     channelId = slackChannelIds[channelName];
```
Am I understanding correctly that you're reassigning the original `channelId` here, which was the channel in which the slash command was issued, to now be the channel that was indicated in the slash command? If so, can you use separate variables for those two separate concepts?

Or are you intending to have the slash command respond the same way to `*` and it would to being invoked without an argument? If so, can you please refactor or add comments? I'm having a hard time following what's going on.

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -396,26 +402,26 @@ export async function handleCommandUnclaimed(
-  if (text && text != '*') {
-    if (text[0] == '#') {
-      text = text.substr(1);
+  if (arg && arg != '*') {
```
What happens if arg is empty string? AFAICT you call `getUnclaimedVoters(null)` similarly to if it were `*` but later it appears the code would handle it as if it were a state.

liewegas replied:
in the empty arg case we show the unclaimed voters for the current channel. new code path is clearer

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -383,11 +383,17 @@ export function prettyTimeInterval(seconds: number): string {
 
-  const slackChannelIds = text
+  var arg = text;
+  if (text && !SlackApiUtil.isMemberOfAdminChannel(userId)) {
```
I'm having hard time following what you intend to happen if arg is an empty string. Can you send a message letting the user know that this slash command is for admins only?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -383,11 +383,17 @@ export function prettyTimeInterval(seconds: number): string {
 export async function handleCommandUnclaimed(
```
I believe there is something specific you need to do in order to ensure that the slash command itself appears (before the response). Did you come across that / does the flow of how this works currently include the slash command itself, so that it's clear to other podmates why the message with the data was sent?

liewegas replied:
Right now they are posted in the channel.  I think instead it should be privately shown to the user, but then the content has to be returned as a response to the request; will follow up with that later after this is merged

tomerovadia commented on src/db_api_util.ts:
```diff
@@ -697,18 +697,24 @@ export async function getThreadsNeedingAttentionFor(
+        , t.user_id
         , EXTRACT(EPOCH FROM now() - updated_at) as age
-        FROM threads t
+        FROM threads t, claims c
```
I believe this crossjoin will duplicate every row in threads by every row in claims which can cause issues and I don't believe is what you're looking for here (e.g. if a vol claims a volunteer multiple times you will end up returning two threads for that thread).

I think what you want is to query volunteer_voter_claims for the latest row with the given volunter_slack_user_id, then join onto that table the needs_attention column from the thread table, which you can then use to filter. I think you want to JOIN on both slack_parent_message_ts and slack_channel_id to make sure you get a one or none match.

liewegas replied:
fixed in latest version

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -494,9 +494,91 @@ export async function handleCommandNeedsAttention(
+        }> - oldest ${prettyTimeInterval(x.maxLastUpdateAge)}`
+    );
+    prefix = 'Voters needing attention by channel';
+  } else if (arg && arg[0] == '@') {
```
Can you put all the arg parsing logic in `command_util`?

tomerovadia commented on src/slack_interaction_handler.ts:
```diff
@@ -528,16 +528,18 @@ export async function handleCommandNeedsAttention(
-    );
+    const vstats = await DbApiUtil.getThreadsNeedingAttentionByVolunteer();
+    for (const x of vstats) {
+      const userName = await SlackApiUtil.getSlackUserName(
```
Can you remove the need for this call by adding slack user name to the DB?

tomerovadia commented on src/slack_api_util.ts:
```diff
@@ -355,6 +359,82 @@ export async function updateSlackChannelNamesAndIdsInRedis(
   }
 }
 
+export async function fetchSlackUserNamesAndIds(): Promise<SlackUserNamesAndIds | null> {
```
Can we write to the DB instead of this?

tomerovadia reviewed (commented):
I'm starting to think we're giving the server and Slack app a role that can be fulfilled far better, more simply and with no risk to production critical operations by a basic data analytics tool. The operations for volunteers are great, but for admin commands, I think having admins log into a tool where they can see overall stats makes much more sense than maintaining so much server logic that requires so much complexity for even basic queries. What do you think?

Also, I've been reviewing this commit-by-commit, which I'm realizing has been a huge mistake. I'm a few from the end but really need to shift gears to tonight's release. If you can let me know when you feel it's ready I can take another look at the whole thing altogether.

liewegas committed:
adjust sql to not leave lobby threads as needs_attention

liewegas committed:
drop archived

liewegas committed:
cleanup

liewegas committed:
rename some loop vars

liewegas committed:
fix slack.md

liewegas committed:
comments

liewegas committed:
handleShowNeedsAttention -> handleShortcutShowNeedsAttention

liewegas committed:
clean up handleCommandUnclaimed

liewegas committed:
clarify with comments

liewegas committed:
tolerate missing username from escaped <@id|name>

From slack:

Escaped usernames may include the plain text username as well, with a bar (|) separator (like this <@U012ABCDEF|worf>).
However, as we are phasing out usernames, your app should only account for the possible presence of that plain text username and separator, but not rely on them.

liewegas committed:
remove more dead code

liewegas committed:
fix Reset Demo on old message w/ no redisData

liewegas committed:
needs-attention: rely on slack links and avoid usernames entirely

liewegas committed:
/needs-attention #channel - fix parsing of escaped arg

liewegas committed:
fix thread creation

liewegas committed:
simplify header above /unclaimed

Otherwise we see #privategroup here, which is confusing

liewegas committed:
prettier

liewegas committed:
Merge remote-tracking branch 'tomerovadia/master' into thread-status

liewegas committed:
fix index

liewegas committed:
add missing exception mask

liewegas committed:
improve naming a bit in the sql

liewegas committed:
use newest claim for each thread

liewegas committed:
refused/spam do not need attention

liewegas committed:
fix thread window

liewegas committed:
create missing threads when we get messages

liewegas committed:
prettier
